### ISSUE TITLE
provide Service-to-Service authorization

### ISSUE THREAD

**Issue by cypherfox** [linkerd/linkerd2#3342]
### What problem are you trying to solve?
Eliminating the need for any authentication and authorization in application micro services. As linkerd2 already has fairly strong identity management infrastructure for automatic TLS, it seems a good starting point.

### How should the problem be solved?
By having the proxy request authorization (authz) from a service (e.g. an OPA wrapper), caching the for a TTL returned with the authz result.

### What do you want to happen? Add any considered drawbacks.
Deny access to the target service if the identity does not comply with a pre-set policy.
For an initial prototype a whitelist of identities per service would be sufficient.

The request to the authz service may cause latency spikes, especially if the service is unavailable.

### Any alternatives you've considered?

Building *good* Authz and Authn code in every micro service. This has never the level of quality that I would aim for.

### How would users interact with this feature?

By configuring the authz service. The method and complexity of it would depend on the type of the authz service.

**Comment by cypherfox** [comment]
I have had a look at the traffic access control part of the SMI spec, as recomended by @grampelberg. The definition is covering my intended use case fine.
(see: https://github.com/deislabs/smi-spec/blob/master/traffic-access-control.md)
While this issue is more focussed on getting authz data to the proxy, being able to define the policy as CRD would be very welcome.

Since we are looking at using other policy engines like OPA, it would be usefull if the SMI CRDs where not *only* way to set authz policy.

**Comment by cypherfox** [comment]
Update to the spec doc: https://github.com/servicemeshinterface/smi-spec/tree/master/apis/traffic-access/

**Comment by jgarces21** [comment]
I think this one is related to #2746

**Comment by cypherfox** [comment]
Hmmm... not as originally written, since Network Polices (NP) and client authentication approach restricting access from two different angles. IMHO The former is more like a machine based (think NFS v3) restriction, while the latter identifies different users (human or machine) and assigns permissions.

I hope SMI is comming soon,  but do not doubt the correctness of @grampelberg's observation on most people only needing NPs

**Comment by glenjamin** [comment]
Is this actually `wontfix`?

I've been trying to find an issue I can subscribe to for future Access Control work, but so far i've only found https://github.com/linkerd/linkerd2/issues/4647 and this one.

**Comment by olix0r** [comment]
This is planned for stable-2.11 (we're currently working on stable-2.10). This has mostly been blocked on extending mTLS to cover all meshed communication. In 2.9 we added mTLS for most TCP traffic; and 2.10 will extend coverage to include server-first protocols, as well as multicluster routing. The server-first mTLS should merge before the new year. We'll update this issue with llinks to our plans for authorization as they are documented. I'd expect this to move forward in January.

**Comment by glenjamin** [comment]
Awesome news - thank you for the update!

**Comment by christianhuening** [comment]
wow! great stuff!

**Comment by erwbgy** [comment]
@olix0r Is the plan to implement SMI Traffic Access Control, support calling an external authz provider (#2901) or something else?

As @cypherfox, @kubaj and others have noted, some of us already have an external authz provider that implements the Envoy External Authorization [API](https://www.envoyproxy.io/docs/envoy/latest/api-v3/service/auth/v3/external_auth.proto#envoy-v3-api-msg-service-auth-v3-checkrequest) (like [opa-envoy-plugin](https://github.com/open-policy-agent/opa-envoy-plugin)) so it would be great if Linkerd could be configured to use it like we can with our ingress controllers.

**Comment by bklaubos** [comment]
Ability for OPA to integrate with something like Okta would be must useful

**Comment by olix0r** [comment]
I'm happy to share our initial proposal/prototype for policy in Linkerd: https://github.com/linkerd/polixy/blob/main/DESIGN.md. I'd like to keep discussions about this proposal in the [discussions for that repo](https://github.com/linkerd/polixy/discussions).

TLDR:

* Two new CRDs: [`Server`][srv] and [`ServerAuthorization`][authz];
  * OPA/GateKeeper/etc may be used to control what types of authorizations may be created.
* Purely connection-oriented for now -- HTTP verbs/URIs/headers are not part of this initial specification. We expect this support to follow.
* We plan to provide an [SMI-compatible adapter][smi] as a Linkerd extension.

These choices have been made so that we can:

* Minimize runtime complexity/overhead.
* Provide a simple solution that can be adopted incrementally.
* Embraces Kubernetes primitives;
* Establishing building blocks that we can reuse for other (non-Authorization) types of server-side configuration.

And, this plan doesn't preclude other solutions in the future.

We're eager for [questions and feedback](https://github.com/linkerd/polixy/discussions)!

[srv]: https://github.com/linkerd/polixy/blob/b448b30944a57dba26279f9466079cc272cdffcb/DESIGN.md#server
[authz]: https://github.com/linkerd/polixy/blob/b448b30944a57dba26279f9466079cc272cdffcb/DESIGN.md#serverauthorization
[smi]: https://github.com/linkerd/polixy/blob/b448b30944a57dba26279f9466079cc272cdffcb/DESIGN.md#why-not-accesssmi-specorg

**Comment by olix0r** [comment]
Linkerd stable-2.11.0 includes support for server authorization policies.
